packages/firestore/src/local/indexeddb_schema.ts

```diff
@@ -582,18 +588,19 @@ export class DbTargetChange {
     /**
      * The snapshot version for this change.
      */
-    public snapshotVersion: DbTimestamp,
+    public snapshotVersion: DbTimestampArray,
     /**
      * The keys of the changed documents in this snapshot.
      */
-    public changes: {
-      added?: EncodedResourcePath[];
-      modified?: EncodedResourcePath[];
-      removed?: EncodedResourcePath[];
-    }
+    public changes: EncodedResourcePath[]
```

Comment:
So...  I think you merged added/modified/removed in response to my feedback. And I gave you that feedback because you don't need to know the type of the change in order to update local views...  but in thinking about this change and the later changes that will use this information, I'm left uncertain...

I guess it boils down to a question (which we may want to discuss offline):  Are you intending to keep `View._syncedDocuments` correctly updated in secondary tabs?  This is not necessary for local view evaluation, and so I think the reasons you might answer yes would be if: a) secondary tabs are responsible for detecting limbo documents and notifying the primary tab to resolve them, or b) we want secondary tabs to have this pre-populated in case they need to take over as primary (but I don't think this will work in general because secondary tabs won't have a View for every active listen across all tabs anyway)

If you are intending to keep it in sync, then I think you may want to preserve the change type here.  If you are not, then I'm wondering if a "Query Change Log" is actually what we want or if we should have a "Remote Document Cache Change Log" instead.  I think this might actually be simpler and may be my preference.  Secondary tabs would only care about keeping their local views correct, which is a consequence of the mutation queue and the remote document cache (well, and whether targets are current, which we need to solve still either way). This would also make sure that secondary clients see when we update remote documents after a write ack (https://github.com/firebase/firebase-js-sdk/blob/a56cbafd564a1e9cad5be366fa58643118745c92/packages/firestore/src/local/local_store.ts#L767) though maybe that isn't strictly necessary (in the cases that matter, we should [eventually at least] get an update from watch as well, which would show up in the query cache change log, I guess].

We can talk more but in general I'm nervous that secondary tabs will be surfacing query results based on remote document cache + mutation queue, but the notification mechanism for changes is via this query cache changelog which is tracking a slightly different thing...

Reply:
As we discussed last week, I have reworked this to use a document change log instead. My main argument against this was churn, and the main argument for this was that this will definitely solve the limbo document problem without any further work. Let me know what you think of the updated PR.

One nice outcome of this is that I no longer have to build up and buffer changes, since the RemoteDocumentBuffer already exists for this purpose.